src/payment/views.py

```diff
@@ -0,0 +1,154 @@
+                ),
+                cancel_url=request.build_absolute_uri(
+                    reverse("v1:payment:subscription-cancel")
+                ),
+                metadata={
+                    "user_id": user.id,
+                    "product_name": product.name,
+                },
+            )
+
+            data = {
+                **request.data,
+                "stripe_subscription_id": checkout_session.id,
+            }
+
+            headers = self.get_success_headers(data)
+            return Response(
+                {
+                    "sessionId": checkout_session.id,
+                    "checkout_url": checkout_session.url,
+                },
+                status=status.HTTP_201_CREATED,
+                headers=headers,
+            )
+        except Exception as e:
+            print("Error during session creation:", e)
+            return Response({"error": str(e)}, status=500)
+
+    def retrieve(self, request, *args, **kwargs):
+        obj = get_object_or_404(self.get_queryset(), user=request.user)
+        serializer = self.get_serializer(obj)
+        return Response(serializer.data)
+
+
+class SubscriptionUserCancelView(APIView):
+    """
+    Handles the cancellation of a subscription by the user.
+
+    This endpoint is used to cancel a subscription by the user.
+    It is called when the user requests to cancel their subscription.
+    """
+
+    def post(self, request, *args, **kwargs):
+        user = request.user
+        subscription = get_object_or_404(Subscription, user=user)
+        stripe.api_key = settings.STRIPE_API_KEY
+
+        try:
+            stripe.Subscription.modify(
+                subscription.stripe_subscription_id, cancel_at_period_end=True
+            )
+
+            subscription.cancel_at_period_end = True
+            subscription.save()
+
+            return Response(
+                {"message": "Subscription cancelled successfully"},
+                status=status.HTTP_200_OK,
+            )
+        except stripe.error.InvalidRequestError as e:
+            return Response(
+                {"error": f"Stripe error: {str(e)}"},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+        except Exception as e:
+            return Response(
+                {"error": f"An error occurred: {str(e)}"},
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            )
+
+
+class SubscriptionSuccessView(APIView):
+    """
+    Handles the redirection after a successful payment.
+
+    This endpoint corresponds to the success_url defined when creating
+    a Stripe checkout session. Stripe redirects the user here when the
+    payment is successful, confirming the subscription action with a
+    200 OK status.
+    """
+
+    def get(self, request, *args, **kwargs):
+        return Response(
+            {"message": "Subscription successful"}, status=status.HTTP_200_OK
+        )
```

Comment:
Does it mean the user gets redirected to an API rather than to a frontend page? Or is this somehow intercepted/handled by the frontend?

Reply:
Yes the user gets redirected there, in a follow up this will be removed once the frontend is implemented.
@issa-diallo would you mind creating a task for the follow up to remove it so we don't keep dead code?